[mlir][bufferization] Simplify helper `potentiallyAliasesMemref`

matthias-springer opened the pull request:
This commit simplifies a helper function in the ownership-based buffer deallocation pass. Fixes a potential double-free (depending on the scheduling of patterns).

maerhart commented on mlir/lib/Dialect/Bufferization/Transforms/BufferDeallocationSimplification.cpp:
```diff
@@ -314,11 +308,13 @@ struct SplitDeallocWhenNotAliasingAnyOther
+    for (int64_t i = 0, e = deallocOp.getMemrefs().size(); i < e; ++i) {
+      Value memref = deallocOp.getMemrefs()[i];
+      Value cond = deallocOp.getConditions()[i];
+      SmallVector<Value> otherMemrefs(deallocOp.getMemrefs());
+      otherMemrefs.erase(otherMemrefs.begin() + i);
```
This probably even fixes a bug when the same memref is contained in `deallocOp.getMemrefs()` multiple times. The old implementation would allow splitting which would lead to a double-free if I understand correctly without actually testing it or looking into it in detail. Do we have a test-case for that?

matthias-springer replied:
I tried to write a test for that. We have another pattern that removes duplicate memrefs (`DeallocRemoveDuplicateDeallocMemrefs`). I think whether the bug is triggered or not depends  on the order in which the patterns are applied. I was not able to trigger the double dealloc in practice. (I'll add the test case back anyway, just to be sure.)

maerhart replied:
Makes sense. It's a bit annoying that MLIR doesn't have the utility to easily test individual patterns :/

matthias-springer committed:
[mlir][bufferization][NFC] Simplify helper `potentiallyAliasesMemref`

This commit simplifies a helper function in the ownership-based buffer deallocation pass.

aartbik commented on mlir/lib/Dialect/Bufferization/Transforms/BufferDeallocationSimplification.cpp:
```diff
@@ -77,18 +77,12 @@ static bool distinctAllocAndBlockArgument(Value v1, Value v2) {
-/// argument indicates whether `memref` may be present in `otherList` which
-/// makes this helper function applicable to situations where we already know
-/// that `memref` is in the list but also when we don't want it in the list.
+/// Checks if `memref` may potentially alias a MemRef in `otherList`. It is
```
This was already there, but I find the "It is often a requirement ..." part a bit awkwardly phrased. 
Can we rephrase this into something like

This test if often needed by optimizations that are only valid when aliasing cannot occur

or something like that (feel free to correct my English please ;-)

matthias-springer replied:
I'm going to rewrite this entire function in a future commit that I haven't sent out for review yet. Another issue here is that we are interested in "same allocation" and not "aliasing". These are slightly different properties. I'm going to update that part of the comment as well.

aartbik reviewed (commented):
LGTM
